### ISSUE TITLE
Show expected and actual renderings for reftests

### ISSUE THREAD

**Issue by foolip** [w3c/wptdashboard#217]
This would have helped me understand https://github.com/w3c/wptdashboard/issues/216

This would require storing a lot of data, but it seems like understanding reftest failures in the dashboard is going to be very challenging otherwise. Even where passing, comparing the pass bitmap in different browsers might reveal something interesting, so we should probably always store it.

**Comment by gsnedders** [comment]
Screenshots are relatively unlikely to change between runs, except in the case of flaky tests, so unique storage shouldn't be *too* bad.

It seems like this would be super-useful as giving some hint about the failure, much like the assertion messages we have for JS tests do today.

**Comment by jgraham** [comment]
We can't always store the bitmap because we don't always generate a bitmap (notably in the Firefox implementation one  of the key optimisations is to only generate a PNG when we get an unexpected failure). But in that case, we put the base64 encoded URL in the logs already, and the [reftest analyzer](https://hg.mozilla.org/mozilla-central/raw-file/tip/layout/tools/reftest/reftest-analyzer.xhtml) exists (that variant requires `--log-tbpl` format input), so the problem here is mostly solved, it's just that the solution isn't exposed well.

**Comment by gsnedders** [comment]
@jgraham Well, I mean, I presume most of this issue is about hooking things up so that the dashboard stores it and serves it from somewhere?

**Comment by jgraham** [comment]
Sure, the dashboard could/should store it. But it could/should store the full run logs for debugging which could/should also contain this data, so I'd fix it that way. You probably can't store those for long, so if you really want something more advanced you need to extract the PNGs and store them in some other way, but that seems like it could be solved later.

**Comment by foolip** [comment]
How big is such a log for a full run? Is that what all results are ultimately extracted from, or is it in addition to other output?

**Comment by jgraham** [comment]
Results are extracted from the logs. Raw logs for a full run are rather large (maybe ~1Gb total, uncompressed), but of course they are incredibly compressible (maybe 24:1 or something, so 42Mb compressed per run).

You can also store the generated text logs, which are smaller. Indeed you probably already are, although if it's the mach ones they currently don't include the screenshots. Switching to `--log-tbpl` will be the fastest way to get the existing reftest analyzer working, since it assumes that input format (I also have a copy that works with raw logs).

For  comparison, if you get failing reftests in gecko, the log looks like [this](https://treeherder.mozilla.org/logviewer.html#?job_id=147705295&repo=mozilla-central&lineNumber=3483) and the "Open analyzer" button takes you to [this page](https://hg.mozilla.org/mozilla-central/raw-file/tip/layout/tools/reftest/reftest-analyzer.xhtml#logurl=https://queue.taskcluster.net/v1/task/VdKQdzKaRECvUvVPLPIFww/runs/0/artifacts/public/logs/live_backing.log&only_show_unexpected=1).

**Comment by foolip** [comment]
Thanks James, good places to start poking if we priorize this. But IIUC, this would not give us images for passes, if we want that we'd need to change some more bits, and probably make the whole run slower?

What's the nature of that optimization, is it passing in a checksum of the image to be tested close to the source, to avoid IPC? If so, perhaps the checksum from previous runs could perhaps be used to not output images that we already have stored...

**Comment by jgraham** [comment]
No, the optimisation is deeper than that. There's an internal canvas for the test and one for the ref, and we avoid `.toDataUrl()` entirely by running the image comparison in C++ code  on the backing data (https://dxr.mozilla.org/mozilla-central/source/dom/interfaces/base/nsIDOMWindowUtils.idl#935).

Of course we can make it  generate more images than that, but always generating the image in passing cases seems like a very diminishing return in usefulness vs only generating it in failing (or, rather, unexpected) cases.

**Comment by foolip** [comment]
OK, so at the very least, if we wanted to have screenshots for passing cases as well, we'd have to do some work to not slow things down by much. Having screenshots in the failing cases is certainly more valuable, so we should start there if we do something.

**Comment by foolip** [comment]
Renamed to track just the results collection part, filed https://github.com/web-platform-tests/wpt.fyi/issues/57 for wpt.fyi work that'd be needed to make use of the results.
